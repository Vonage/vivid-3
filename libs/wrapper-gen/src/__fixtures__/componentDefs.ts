import { ComponentDef } from '../metadata/ComponentDef';

export const minimalComponent: ComponentDef = {
	name: 'example-component',
	className: 'ExampleComponent',
	wrappedClassName: 'VExampleComponent',
	vividModulePath:
		'libs/components/src/lib/example-component/example-component.ts',
	registerFunctionName: 'registerExampleComponent',
	attributes: [],
	events: [],
	vueModels: [],
	methods: [],
	slots: [],
	localTypeDefs: {},
};

export const exampleComponent: ComponentDef = {
	name: 'example-component',
	className: 'ExampleComponent',
	wrappedClassName: 'VExampleComponent',
	vividModulePath:
		'libs/components/src/lib/example-component/example-component.ts',
	registerFunctionName: 'registerExampleComponent',
	description: 'This is an example component',
	attributes: [
		{
			name: 'example-attribute',
			description: 'This is an example attribute',
			forwardTo: {
				type: 'attribute',
				name: 'example-attribute',
			},
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
		{
			name: 'type-union',
			forwardTo: {
				type: 'attribute',
				name: 'type-union',
			},
			type: [
				{ text: 'string', vuePropType: 'String' },
				{ text: 'number', vuePropType: 'Number' },
			],
		},
		{
			name: 'imported-type',
			forwardTo: {
				type: 'property',
				name: 'importedType',
			},
			type: [
				{
					text: 'Imported',
					vuePropType: 'Object',
					importFromModule: 'module',
				},
			],
		},
		{
			name: 'value',
			forwardTo: {
				type: 'attribute',
				name: 'value',
			},
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
		{
			name: 'start',
			forwardTo: {
				type: 'attribute',
				name: 'start',
			},
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
		{
			name: 'forced-dom-prop',
			forwardTo: {
				type: 'property',
				name: 'forcedDomProp',
			},
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
		{
			name: 'boolean-attribute',
			forwardTo: {
				type: 'attribute',
				name: 'boolean-attribute',
				boolean: true,
			},
			type: [
				{
					text: 'boolean',
					vuePropType: 'Boolean',
				},
			],
		},
	],
	events: [
		{
			name: 'example-event',
			description: 'This is an example event',
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
		{
			name: 'input',
			type: [
				{
					text: 'Event',
					vuePropType: 'Event',
				},
			],
		},
		{
			name: 'input:start',
			type: [
				{
					text: 'CustomEvent<{a: string | number}>',
					vuePropType: 'CustomEvent',
				},
			],
		},
	],
	vueModels: [
		{
			name: 'modelValue',
			attributeName: 'value',
			eventNames: ['input'],
			valueMapping: '(event.target as any).value',
		},
		{
			name: 'sameEvent',
			attributeName: 'example-attribute',
			eventNames: ['input'],
			valueMapping: '(event.target as any).exampleAttribute',
		},
		{
			name: 'start',
			attributeName: 'start',
			eventNames: ['input:start'],
			valueMapping: '(event.target as any).start',
		},
	],
	methods: [
		{
			name: 'exampleMethod',
			description: 'This is an example method',
			args: [
				{
					name: 'exampleArg',
					type: [
						{
							text: 'string',
							vuePropType: 'String',
						},
					],
				},
			],
			returnType: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
	],
	slots: [
		{
			name: 'example-slot',
			description: 'This is an example slot',
		},
		{
			name: 'no-description',
		},
	],
	localTypeDefs: {},
};
