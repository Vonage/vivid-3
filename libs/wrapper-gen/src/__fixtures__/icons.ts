import { IconsManifest } from '@vonage/vwc-consts';

export const iconsManifestStub: IconsManifest = [
	{
		id: '1-icon-1',
		keyword: ['icon-1'],
		tag: ['style_weight_solid', 'style_color_single', 'category_general'],
	},
	{
		id: 'icon-2',
		keyword: ['icon-2'],
		tag: ['style_weight_regular', 'style_color_multi', 'category_general'],
	},
];
