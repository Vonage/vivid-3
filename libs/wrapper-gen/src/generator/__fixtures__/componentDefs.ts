import { ComponentDef } from '../ComponentDef';

export const minimalComponent: ComponentDef = {
	name: 'example-component',
	className: 'ExampleComponent',
	wrappedClassName: 'VExampleComponent',
	vividModulePath:
		'libs/components/src/lib/example-component/example-component.ts',
	registerFunctionName: 'registerExampleComponent',
	attributes: [],
	events: [],
	vueModels: [],
	methods: [],
	slots: [],
	localTypeDefs: {},
};

export const exampleComponent: ComponentDef = {
	name: 'example-component',
	className: 'ExampleComponent',
	wrappedClassName: 'VExampleComponent',
	vividModulePath:
		'libs/components/src/lib/example-component/example-component.ts',
	registerFunctionName: 'registerExampleComponent',
	description: 'This is an example component',
	attributes: [
		{
			name: 'example-attribute',
			description: 'This is an example attribute',
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
		{
			name: 'type-union',
			type: [
				{ text: 'string', vuePropType: 'String' },
				{ text: 'number', vuePropType: 'Number' },
			],
		},
		{
			name: 'imported-type',
			type: [
				{
					text: 'Imported',
					vuePropType: 'Object',
					importFromModule: 'module',
				},
			],
		},
		{
			name: 'value',
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
		{
			name: 'start',
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
		{
			name: 'forced-dom-prop',
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
			forceDomProp: true,
		},
	],
	events: [
		{
			name: 'example-event',
			description: 'This is an example event',
			type: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
		{
			name: 'input',
			type: [
				{
					text: 'Event',
					vuePropType: 'Event',
				},
			],
		},
		{
			name: 'input:start',
			type: [
				{
					text: 'CustomEvent<{a: string | number}>',
					vuePropType: 'CustomEvent',
				},
			],
		},
	],
	vueModels: [
		{
			name: 'modelValue',
			attributeName: 'value',
			eventName: 'input',
			valueMapping: '(event.target as any).value',
		},
		{
			name: 'start',
			attributeName: 'start',
			eventName: 'input:start',
			valueMapping: '(event.target as any).start',
		},
	],
	methods: [
		{
			name: 'exampleMethod',
			description: 'This is an example method',
			args: [
				{
					name: 'exampleArg',
					type: [
						{
							text: 'string',
							vuePropType: 'String',
						},
					],
				},
			],
			returnType: [
				{
					text: 'string',
					vuePropType: 'String',
				},
			],
		},
	],
	slots: [
		{
			name: 'example-slot',
			description: 'This is an example slot',
		},
		{
			name: 'no-description',
		},
	],
	localTypeDefs: {},
};
