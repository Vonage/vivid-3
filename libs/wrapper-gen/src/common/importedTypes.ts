import { fetchIconsManifest, iconTypeFromManifest } from './icons';
import { Metadata } from './metadata';
import { makeTypeResolver } from './types';

export const makeImportedTypesResolver = async (metadata: Metadata) =>
	makeTypeResolver({
		IconId: iconTypeFromManifest(
			await fetchIconsManifest(metadata.iconsManifestUrl)
		),
	});
