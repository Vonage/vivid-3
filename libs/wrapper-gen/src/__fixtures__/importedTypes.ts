import { makeTypeResolver } from '../common/types';
import { iconTypeFromManifest } from '../common/icons';
import { iconsManifestStub } from './icons';

export const importedTypesResolverStub = makeTypeResolver({
	IconId: iconTypeFromManifest(iconsManifestStub),
});
