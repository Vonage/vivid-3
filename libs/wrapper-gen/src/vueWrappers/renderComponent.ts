import { ComponentDef } from '../common/ComponentDef';
import { kebabToCamel, kebabToPascal } from '../utils/casing';
import { parseTypeStr, TypeResolver, TypeStr } from '../common/types';
import { getExportedClassName } from '../metadata/vividPackage';
import { vuePropTypes } from './vuePropTypes';
import { wrappedComponentName } from './name';
import { determinePropForwarding } from './propForwarding';
import { Import, importsForTypes, renderImports } from './imports';

const renderJsDoc = (description?: string, type?: TypeStr) => {
	if (!description && !type) return '';

	const renderedDescription = description
		? `\n${description
				.split('\n')
				.map((line) => ` * ${line}`)
				.join('\n')}`
		: '';

	const renderedType = type ? `\n * @type {${type}}` : '';

	return `/**${renderedDescription}${renderedType}
 */`;
};

export const renderComponent = (
	componentDef: ComponentDef,
	importedTypesResolver: TypeResolver,
	isVue3Stub = false
) => {
	const vueModule = isVue3Stub ? 'vue3' : 'vue';

	const imports: Import[] = [
		{ name: 'defineComponent', fromModule: vueModule },
		{ name: 'ref', fromModule: vueModule },
		{ name: 'h', fromModule: vueModule },
		{ name: 'isVue2', fromModule: '../../utils/vue' },
		{ name: 'handleVue3Props', fromModule: '../../utils/ssr' },
		{ name: 'VNodeData', fromModule: vueModule },
		{ name: componentDef.registerFunctionName, fromModule: '@vonage/vivid' },
		{ name: 'registerComponent', fromModule: '../../utils/register' },
	];

	// Filter out props that are overshadowed by v-model name
	const props = componentDef.props.filter(
		({ name }) =>
			!componentDef.vueModels.some(
				(model) => model.name === name && model.propName !== name
			)
	);

	// Find v-models and their corresponding prop and event
	const vueModels = componentDef.vueModels.map((model) => {
		const prop = componentDef.props.find(
			(prop) => prop.name === model.propName
		);
		if (!prop) throw new Error(`v-model prop not found: ${model.propName}`);
		for (const eventName of model.eventNames) {
			const event = componentDef.events.find((e) => e.name === eventName);
			if (!event) throw new Error(`v-model event not found: ${eventName}`);
		}

		return {
			...model,
			prop,
		};
	});

	const vueModelEvents = vueModels.map((vueModel) => ({
		name: `update:${vueModel.name}`,
		description: `Fires when the ${vueModel.name} value changes`,
		type: vueModel.prop.type,
	}));

	if (props.length > 0) {
		imports.push({ name: 'PropType', fromModule: vueModule });
	}

	const typeImports: Import[] = [
		{
			name: getExportedClassName(componentDef.name),
			fromModule: '@vonage/vivid',
		},
	];

	// Import referenced types
	const referencesTypes = [
		...props.map((prop) => prop.type),
		...componentDef.events.map((event) => event.type),
		...componentDef.methods.flatMap((method) =>
			method.args.map((arg) => arg.type)
		),
	].flatMap(parseTypeStr);
	imports.push(...importsForTypes(referencesTypes));

	if (isVue3Stub) {
		typeImports.push({ name: 'SlotsType', fromModule: vueModule });
	}

	/**
	 * Forward all props to their respective attribute / dom property.
	 * In Vue 2, attributes and properties are under separate keys.
	 * In Vue 3, we need to prefix attributes with '^' and properties with '.' to differentiate them.
	 */
	const renderProps = (
		props: ComponentDef['props'],
		syntax: 'vue2-attrs' | 'vue2-domProps' | 'vue3'
	) => {
		const propsSrc = props
			.flatMap((prop) => {
				const forwardTo = determinePropForwarding(
					prop,
					parseTypeStr(importedTypesResolver(prop.type))
				);

				if (
					(syntax === 'vue2-attrs' && forwardTo.type === 'property') ||
					(syntax === 'vue2-domProps' && forwardTo.type === 'attribute')
				) {
					return [];
				}

				const vueModel = componentDef.vueModels.find(
					(model) => model.propName === prop.name
				);

				let valueToUse = `this.${prop.name}`;
				if (vueModel && vueModel.name !== prop.name) {
					// If there is a v-model, we will prefer the v-model value
					valueToUse = `this.${vueModel.name} ?? ${valueToUse}`;
				}

				const nameToUse =
					syntax === 'vue3'
						? forwardTo.type === 'attribute'
							? `^${forwardTo.name}`
							: `.${forwardTo.name}`
						: forwardTo.name;

				// Vue 2 and 3 differ in how they handle boolean attributes
				// Remove false attributes in Vue 3 to make it behave like Vue 2
				const booleanFilter =
					forwardTo.type === 'attribute' &&
					forwardTo.boolean &&
					syntax === 'vue3'
						? ` && (${valueToUse}) !== false`
						: '';

				// Vue requires us to filter out undefined properties before passing them into the h function
				const filter = `(${valueToUse}) !== undefined${booleanFilter}`;

				return [`...(${filter} ? {'${nameToUse}': ${valueToUse} } : {})`];
			})
			.join(',');
		if (syntax === 'vue3') {
			return `...handleVue3Props({${propsSrc}})`;
		}
		return propsSrc;
	};

	const propsV3Src = renderProps(props, 'vue3');

	const attrsV2Src = renderProps(props, 'vue2-attrs');
	const domPropsV2Src = renderProps(props, 'vue2-domProps');

	const renderEventType = (type: TypeStr, isVueModelEvent: boolean): string => {
		// Event type should be a single type like `CustomEvent<undefined>`
		if (parseTypeStr(type).length > 1) {
			throw new Error('Multiple event types not supported');
		}

		if (isVueModelEvent) {
			return type; // Vue model events use the prop's type, e.g. a `update:modelValue` will have type `string`
		}

		// The event `currentTarget` will always be the host component. Therefore, type `currentTarget` accordingly to make
		// it easier to use for consumers.
		// Originally we typed `target` instead, but this is not the case if the event bubbles from the light DOM, e.g. an
		// input event bubbling from a select slotted into a text-field. We will remove the inaccurate typing in a future
		// major version.
		return `${type} & {
			/**
			 * @deprecated Target may not refer to component in some cases. Use currentTarget instead.
			 */
			target: ${getExportedClassName(componentDef.name)},
			currentTarget: ${getExportedClassName(componentDef.name)}
		}`;
	};

	/**
	 * All events should be forwarded
	 */
	const eventsSrc = componentDef.events
		.map(({ name, type }) => {
			const eventVueModels = vueModels.filter((model) =>
				model.eventNames.includes(name)
			);
			return `'${name}': (event: ${renderEventType(type, false)}) => {
          ${eventVueModels
						.map(
							(vueModel) =>
								`this.$emit('update:${vueModel.name}', ${vueModel.valueMapping});`
						)
						.join('\n')}
          this.$emit('${name}', event);
        }`;
		})
		.join(',');

	const eventsV3Src = componentDef.events
		.map(({ name, type }) => {
			const eventVueModels = vueModels.filter((model) =>
				model.eventNames.includes(name)
			);
			return `'on${kebabToPascal(name)}': (event: ${renderEventType(
				type,
				false
			)}) => {
					 ${eventVueModels
							.map(
								(vueModel) =>
									`this.$emit('update:${vueModel.name}', ${vueModel.valueMapping});`
							)
							.join('\n')}
          this.$emit('${name}', event);
        }`;
		})
		.join(',');

	const slotsDeclarationSrc = isVue3Stub
		? `slots: Object as SlotsType<${
				componentDef.slots.length
					? `{
		${componentDef.slots
			.map(
				(slot) => `
		${renderJsDoc(slot.description)}
		"${slot.name}": Record<string, never>`
			)
			.join('\n')}
	}`
					: 'Record<string, never>'
		  }>,`
		: '';

	const renderNamedSlots = (syntax: 'vue2' | 'vue3') =>
		componentDef.slots
			.filter((slot) => slot.name !== 'default')
			.map(
				(slot) => `// @slot ${slot.name} ${slot.description ?? ''}
      ${syntax === 'vue2' ? 'handleNamedSlotV2' : 'handleNamedSlotV3'}('${
					slot.name
				}', this.$slots['${slot.name}']${syntax === 'vue3' ? ' as any' : ''})`
			)
			.join(',');

	const namedSlotsV2Src = renderNamedSlots('vue2');
	const namedSlotsV3Src = renderNamedSlots('vue3');

	if (namedSlotsV2Src) {
		imports.push({
			name: 'handleNamedSlotV2',
			fromModule: '../../utils/slots',
		});
	}

	if (namedSlotsV3Src) {
		imports.push({
			name: 'handleNamedSlotV3',
			fromModule: '../../utils/slots',
		});
	}

	const renderPropType = (type: TypeStr): string => {
		const propTypes = vuePropTypes(importedTypesResolver(type));
		return propTypes.length > 1 ? `[${propTypes.join(', ')}]` : propTypes[0];
	};

	/**
	 * Declare props.
	 * Note: All props are optional. Setting default to undefined, otherwise Vue 3 will default boolean props to false.
	 * myProp: {type: [String, Number] as PropType<string | number>, default: undefined},
	 * Note: When there are no props, props key needs to be omitted or the typings will break in Vue 3.
	 */
	const propDefinitionsSrc = props.length
		? `props: { ${props
				.flatMap((prop) => {
					const vueModel = componentDef.vueModels.find(
						(model) => model.propName === prop.name
					);

					return [
						prop,
						...(vueModel && vueModel.name !== prop.name
							? [
									{
										...prop,
										name: vueModel.name,
									},
							  ]
							: []),
					];
				})
				.map(({ name, description, type }) => {
					const propName = kebabToCamel(name);
					return `${renderJsDoc(description)}
        ${propName}: {type: ${renderPropType(
						type
					)} as PropType<${type}>, default: undefined}`;
				})
				.join(',\n')}
			},`
		: '';

	const renderEventDeclaration = ({
		name,
		description,
		type,
	}: ComponentDef['events'][number]) =>
		`
					${renderJsDoc(description, type)}
					'${name}'`;

	const renderEventDeclarationV3 =
		(isVueModelEvent: boolean) =>
		({ name, description, type }: ComponentDef['events'][number]) =>
			`
				${renderJsDoc(description, type)}
				['${name}'](event: ${renderEventType(type, isVueModelEvent)}) { return true }`;

	// Declare events
	const eventDefinitionsSrc = isVue3Stub
		? `{
			${[
				...componentDef.events.map(renderEventDeclarationV3(false)),
				...vueModelEvents.map(renderEventDeclarationV3(true)),
			].join(',\n')}}`
		: `[
			${[...componentDef.events, ...vueModelEvents]
				.map(renderEventDeclaration)
				.join(',\n')}]`;

	// For vue2, we rename v-model prop and event to the vue3 default names
	const vue2VModelSrc = vueModels.some((model) => model.name === 'modelValue')
		? `
  model: isVue2 ? {
    prop: 'modelValue',
    event: 'update:modelValue'
  } : undefined,`
		: '';

	// No need to render body for vue3 stubs, would run into type errors otherwise
	const renderMethodBody = isVue3Stub
		? 'return null;'
		: `
    if (isVue2) {
        return h(this.componentName, {
            ref: 'element',
            attrs: { ${attrsV2Src} },
            class: 'vvd-component',
            ${domPropsV2Src ? `domProps: { ${domPropsV2Src} },` : ''}
            on: { ${eventsSrc} },
        }, [
            this.$slots.default,
            ${namedSlotsV2Src}
        ]);
      }
      // @ts-ignore
      return h(this.componentName, {
          ref: 'element',
          class: 'vvd-component',
          ${[propsV3Src, eventsV3Src].filter(Boolean).join(',')}
      } as unknown as VNodeData, [
          // @ts-ignore
          this.$slots.default && this.$slots.default(),
          ${namedSlotsV3Src}
      ]);
    `;

	/**
	 * Forward methods like this:
	 * focus: (arg: string): void => {
	 *   this.element?.focus(arg);
	 * },
	 */
	const methodDefinitionsSrc = componentDef.methods
		.map(
			(method) =>
				`
        ${renderJsDoc(method.description)}
        ${method.name}(${method.args
					.map((a) => `${a.name}: ${a.type}`)
					.join(', ')}): ${method.returnType}{ return (this.element as any)?.${
					method.name
				}(${method.args.map((a) => a.name).join(', ')}); }`
		)
		.join(',\n');

	return `
${renderImports(imports)}
${renderImports(typeImports, true)}


${renderJsDoc(componentDef.description)}
export default defineComponent({
  name: '${wrappedComponentName(componentDef)}',
  ${vue2VModelSrc}
  ${propDefinitionsSrc}
  emits: ${eventDefinitionsSrc},
  methods: {
  	${methodDefinitionsSrc}
	},
	${slotsDeclarationSrc}
  setup(props, ctx) {
    const componentName = registerComponent('${componentDef.name}', ${
		componentDef.registerFunctionName
	});

    const element = ref<${getExportedClassName(
			componentDef.name
		)} | null>(null);

    return { componentName, element };
  },
  render() {
    ${renderMethodBody}
  },
});
`;
};
