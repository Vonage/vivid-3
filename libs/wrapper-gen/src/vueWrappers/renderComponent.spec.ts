import { renderComponent } from './renderComponent';
import {
	exampleComponent,
	minimalComponent,
} from '../__fixtures__/componentDefs';
import { importedTypesResolverStub } from '../__fixtures__/importedTypes';

describe('renderComponent', () => {
	it('should render a component', () => {
		expect(
			renderComponent(exampleComponent, importedTypesResolverStub)
		).toMatchSnapshot();
	});

	it('should render a component as a Vue 3 stub', () => {
		expect(
			renderComponent(exampleComponent, importedTypesResolverStub, true)
		).toMatchSnapshot();
	});

	it('should render a minimal component', () => {
		expect(
			renderComponent(minimalComponent, importedTypesResolverStub)
		).toMatchSnapshot();
	});

	it('should render minimal component as a Vue 3 stub', () => {
		expect(
			renderComponent(minimalComponent, importedTypesResolverStub, true)
		).toMatchSnapshot();
	});

	it('should throw an error if the event for a Vue model is not found', () => {
		expect(() =>
			renderComponent(
				{
					...minimalComponent,
					props: [
						{
							name: 'value',
							type: 'string',
							attributeName: 'attribute',
							propertyName: 'value',
						},
					],
					vueModels: [
						{
							name: 'modelValue',
							propName: 'value',
							eventNames: ['not-found'],
							valueMapping: '(event.target as any).value',
						},
					],
				},
				importedTypesResolverStub
			)
		).toThrow('v-model event not found');
	});

	it('should throw an error if the prop for a Vue model is not found', () => {
		expect(() =>
			renderComponent(
				{
					...minimalComponent,
					events: [
						{
							name: 'input',
							type: 'Event',
						},
					],
					vueModels: [
						{
							name: 'modelValue',
							propName: 'not-found',
							eventNames: ['input'],
							valueMapping: '(event.target as any).value',
						},
					],
				},
				importedTypesResolverStub
			)
		).toThrow('v-model prop not found');
	});

	it('should throw an error if the type of an event is a union', () => {
		expect(() =>
			renderComponent(
				{
					...minimalComponent,
					events: [
						{
							name: 'input',
							type: 'Event | CustomEvent',
						},
					],
				},
				importedTypesResolverStub,
				true
			)
		).toThrow('Multiple event types not supported');
	});
});
