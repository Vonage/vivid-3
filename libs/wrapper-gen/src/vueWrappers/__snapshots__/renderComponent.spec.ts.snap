// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renderComponent > should render a component 1`] = `
"
import { defineComponent, ref, h, VNodeData, PropType } from 'vue';
import { isVue2 } from '../../utils/vue';
import { handleVue3Props } from '../../utils/ssr';
import { registerExampleComponent } from '@vonage/vivid';
import { registerComponent } from '../../utils/register';
import { IconId } from '../icons';
import { handleNamedSlot } from '../../utils/slots';
import type { ExampleComponent } from '@vonage/vivid/lib/example-component/example-component';


/**
 * This is an example component
 */
export default defineComponent({
  name: 'VExampleComponent',
  
  model: isVue2 ? {
    prop: 'modelValue',
    event: 'update:modelValue'
  } : undefined,
  props: {
    /**
 * This is an example prop
 */
        exampleProp: {type: String as PropType<string>, default: undefined},
/**
 * This is an example prop
 */
        sameEvent: {type: String as PropType<string>, default: undefined},

        typeUnion: {type: [String, Number] as PropType<string | number>, default: undefined},

        iconType: {type: String as PropType<IconId>, default: undefined},

        value: {type: String as PropType<string>, default: undefined},

        modelValue: {type: String as PropType<string>, default: undefined},

        start: {type: String as PropType<string>, default: undefined},

        forcedDomProp: {type: [String, HTMLElement] as PropType<string | HTMLElement>, default: undefined},

        booleanProp: {type: Boolean as PropType<boolean>, default: undefined}
  },
  emits: [
			
						/**
 * This is an example event
 * @type {string}
 */
							'example-event',

						/**
 * @type {Event}
 */
							'input',

						/**
 * @type {CustomEvent<{a: string | number}>}
 */
							'input:start',

						/**
 * Fires when the modelValue value changes
 * @type {string}
 */
							'update:modelValue',

						/**
 * Fires when the sameEvent value changes
 * @type {string}
 */
							'update:sameEvent',

						/**
 * Fires when the start value changes
 * @type {string}
 */
							'update:start'],
  methods: {
  	
        /**
 * This is an example method
 */
        exampleMethod(exampleArg: string): string{ return (this.element as any)?.exampleMethod(exampleArg); }
	},
	
  setup(props, ctx) {
    const componentName = registerComponent('example-component', registerExampleComponent);

    const element = ref<ExampleComponent | null>(null);

    return { componentName, element };
  },
  render() {
    
    if (isVue2) {
        return h(this.componentName, {
            ref: 'element',
            attrs: { ...((this.sameEvent ?? this.exampleProp) !== undefined ? {'example-prop': this.sameEvent ?? this.exampleProp } : {}),...((this.typeUnion) !== undefined ? {'type-union': this.typeUnion } : {}),...((this.iconType) !== undefined ? {'icon-type': this.iconType } : {}),...((this.modelValue ?? this.value) !== undefined ? {'current-value': this.modelValue ?? this.value } : {}),...((this.start) !== undefined ? {'start': this.start } : {}),...((this.booleanProp) !== undefined ? {'boolean-prop': this.booleanProp } : {}) },
            class: 'vvd-component',
            domProps: { ...((this.forcedDomProp) !== undefined ? {'forcedDomProp': this.forcedDomProp } : {}) },
            on: { 'example-event': (event: Event) => {
          
          this.$emit('example-event', event);
        },'input': (event: Event) => {
          this.$emit('update:modelValue', (event.target as any).value);
this.$emit('update:sameEvent', (event.target as any).exampleProp);
          this.$emit('input', event);
        },'input:start': (event: Event) => {
          this.$emit('update:start', (event.target as any).start);
          this.$emit('input:start', event);
        } },
        }, [
            this.$slots.default,
            // @slot example-slot This is an example slot
      handleNamedSlot('example-slot', this.$slots['example-slot']),// @slot no-description 
      handleNamedSlot('no-description', this.$slots['no-description'])
        ]);
      }
      // @ts-ignore
      return h(this.componentName, {
          ref: 'element',
          class: 'vvd-component',
          ...handleVue3Props({...((this.sameEvent ?? this.exampleProp) !== undefined ? {'^example-prop': this.sameEvent ?? this.exampleProp } : {}),...((this.typeUnion) !== undefined ? {'^type-union': this.typeUnion } : {}),...((this.iconType) !== undefined ? {'^icon-type': this.iconType } : {}),...((this.modelValue ?? this.value) !== undefined ? {'^current-value': this.modelValue ?? this.value } : {}),...((this.start) !== undefined ? {'^start': this.start } : {}),...((this.forcedDomProp) !== undefined ? {'.forcedDomProp': this.forcedDomProp } : {}),...((this.booleanProp) !== undefined && (this.booleanProp) !== false ? {'^boolean-prop': this.booleanProp } : {})}),'onExampleEvent': (event: Event) => {
					 
          this.$emit('example-event', event);
        },'onInput': (event: Event) => {
					 this.$emit('update:modelValue', (event.target as any).value);
this.$emit('update:sameEvent', (event.target as any).exampleProp);
          this.$emit('input', event);
        },'onInput:start': (event: Event) => {
					 this.$emit('update:start', (event.target as any).start);
          this.$emit('input:start', event);
        }
      } as unknown as VNodeData, [
          // @ts-ignore
          this.$slots.default && this.$slots.default(),
          // @slot example-slot This is an example slot
      handleNamedSlot('example-slot', this.$slots['example-slot']),// @slot no-description 
      handleNamedSlot('no-description', this.$slots['no-description'])
      ]);
    
  },
});
"
`;

exports[`renderComponent > should render a component as a Vue 3 stub 1`] = `
"
import { defineComponent, ref, h, VNodeData, PropType } from 'vue3';
import { isVue2 } from '../../utils/vue';
import { handleVue3Props } from '../../utils/ssr';
import { registerExampleComponent } from '@vonage/vivid';
import { registerComponent } from '../../utils/register';
import { IconId } from '../icons';
import { handleNamedSlot } from '../../utils/slots';
import type { ExampleComponent } from '@vonage/vivid/lib/example-component/example-component';
import type { SlotsType } from 'vue3';


/**
 * This is an example component
 */
export default defineComponent({
  name: 'VExampleComponent',
  
  model: isVue2 ? {
    prop: 'modelValue',
    event: 'update:modelValue'
  } : undefined,
  props: {
    /**
 * This is an example prop
 */
        exampleProp: {type: String as PropType<string>, default: undefined},
/**
 * This is an example prop
 */
        sameEvent: {type: String as PropType<string>, default: undefined},

        typeUnion: {type: [String, Number] as PropType<string | number>, default: undefined},

        iconType: {type: String as PropType<IconId>, default: undefined},

        value: {type: String as PropType<string>, default: undefined},

        modelValue: {type: String as PropType<string>, default: undefined},

        start: {type: String as PropType<string>, default: undefined},

        forcedDomProp: {type: [String, HTMLElement] as PropType<string | HTMLElement>, default: undefined},

        booleanProp: {type: Boolean as PropType<boolean>, default: undefined}
  },
  emits: {
			
						/**
 * This is an example event
 * @type {string}
 */
						['example-event'](event: string & { target: ExampleComponent}) { return true },

						/**
 * @type {Event}
 */
						['input'](event: Event & { target: ExampleComponent}) { return true },

						/**
 * @type {CustomEvent<{a: string | number}>}
 */
						['input:start'](event: CustomEvent<{a: string | number}> & { target: ExampleComponent}) { return true },

						/**
 * Fires when the modelValue value changes
 * @type {string}
 */
						['update:modelValue'](event: string & { target: ExampleComponent}) { return true },

						/**
 * Fires when the sameEvent value changes
 * @type {string}
 */
						['update:sameEvent'](event: string & { target: ExampleComponent}) { return true },

						/**
 * Fires when the start value changes
 * @type {string}
 */
						['update:start'](event: string & { target: ExampleComponent}) { return true }},
  methods: {
  	
        /**
 * This is an example method
 */
        exampleMethod(exampleArg: string): string{ return (this.element as any)?.exampleMethod(exampleArg); }
	},
	slots: Object as SlotsType<{
		
		/**
 * This is an example slot
 */
		"example-slot": Record<string, never>

		
		"no-description": Record<string, never>
	}>,
  setup(props, ctx) {
    const componentName = registerComponent('example-component', registerExampleComponent);

    const element = ref<ExampleComponent | null>(null);

    return { componentName, element };
  },
  render() {
    return null;
  },
});
"
`;

exports[`renderComponent > should render a minimal component 1`] = `
"
import { defineComponent, ref, h, VNodeData } from 'vue';
import { isVue2 } from '../../utils/vue';
import { handleVue3Props } from '../../utils/ssr';
import { registerExampleComponent } from '@vonage/vivid';
import { registerComponent } from '../../utils/register';
import type { ExampleComponent } from '@vonage/vivid/lib/example-component/example-component';



export default defineComponent({
  name: 'VExampleComponent',
  
  props: {
    
  },
  emits: [
			],
  methods: {
  	
	},
	
  setup(props, ctx) {
    const componentName = registerComponent('example-component', registerExampleComponent);

    const element = ref<ExampleComponent | null>(null);

    return { componentName, element };
  },
  render() {
    
    if (isVue2) {
        return h(this.componentName, {
            ref: 'element',
            attrs: {  },
            class: 'vvd-component',
            
            on: {  },
        }, [
            this.$slots.default,
            
        ]);
      }
      // @ts-ignore
      return h(this.componentName, {
          ref: 'element',
          class: 'vvd-component',
          ...handleVue3Props({})
      } as unknown as VNodeData, [
          // @ts-ignore
          this.$slots.default && this.$slots.default(),
          
      ]);
    
  },
});
"
`;

exports[`renderComponent > should render minimal component as a Vue 3 stub 1`] = `
"
import { defineComponent, ref, h, VNodeData } from 'vue3';
import { isVue2 } from '../../utils/vue';
import { handleVue3Props } from '../../utils/ssr';
import { registerExampleComponent } from '@vonage/vivid';
import { registerComponent } from '../../utils/register';
import type { ExampleComponent } from '@vonage/vivid/lib/example-component/example-component';
import type { SlotsType } from 'vue3';



export default defineComponent({
  name: 'VExampleComponent',
  
  props: {
    
  },
  emits: {
			},
  methods: {
  	
	},
	slots: Object as SlotsType<Record<string, never>>,
  setup(props, ctx) {
    const componentName = registerComponent('example-component', registerExampleComponent);

    const element = ref<ExampleComponent | null>(null);

    return { componentName, element };
  },
  render() {
    return null;
  },
});
"
`;
